dali/operators/geometry/transform_base_op.h

```diff
@@ -0,0 +1,185 @@
+// Copyright (c) 2020, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_OPERATORS_GEOMETRY_TRANSFORM_BASE_OP_H_
+#define DALI_OPERATORS_GEOMETRY_TRANSFORM_BASE_OP_H_
+
+#include <vector>
+#include "dali/core/format.h"
+#include "dali/core/geom/mat.h"
+#include "dali/core/static_switch.h"
+#include "dali/kernels/kernel_manager.h"
+#include "dali/pipeline/data/types.h"
+#include "dali/pipeline/operator/op_spec.h"
+#include "dali/pipeline/workspace/workspace.h"
+#include "dali/pipeline/operator/operator.h"
+
+#define TRANSFORM_INPUT_TYPES (float)
+
+namespace dali {
+
+template <int... values>
+using dims = std::integer_sequence<int, values...>;
+
+template <typename T, int ndim>
+using affine_mat_t = mat<ndim+1, ndim+1, T>;
+
+/**
+ * @brief Base CRTP class for affine transform generators.
+ * The matrix definition comes from the actual TransformImpl implementation.
+ * As with any CRTP-based system, any non-private method can be shadowed by the TransformImpl class.
+ */
+template <typename Backend, typename TransformImpl>
+class TransformBaseOp : public Operator<Backend> {
+ public:
+  explicit TransformBaseOp(const OpSpec &spec)
+      : Operator<Backend>(spec),
+        spec_(spec),
+        nsamples_(spec.GetArgument<int>("batch_size")) {
+  }
+
+  bool CanInferOutputs() const override { return true; }
+
+  TransformImpl &This() noexcept { return static_cast<TransformImpl&>(*this); }
+  const TransformImpl &This() const noexcept { return static_cast<const TransformImpl&>(*this); }
+
+ protected:
+  void CheckInputDimensionality(const workspace_t<CPUBackend> &ws) {
```

Comment:
I'd avoid the name "dimensionality" - especially as a standalone word. Here I'd use
```suggestion
  void CheckInputShape(const workspace_t<CPUBackend> &ws) {
```